lib/gc/Transforms/AnyTilableFusion.cpp

```diff
@@ -0,0 +1,717 @@
+      targetInnerTileSizes.resize(dimAndTileMapping.size());
+      for (const auto &dimAndTile : dimAndTileMapping) {
+        targetInnerTileSizes[dimAndTile.first] = dimAndTile.second;
+      }
+    } else // tileSize comes from OpOperand
+        if (std::is_same<T, OffsetSizeAndStrideOpInterface>::value) {
+      targetInnerTileSizes = unPackOp.getInnerTiles();
+      targetTileSizes =
+          llvm::to_vector(tileSizes.take_back(targetInnerTileSizes.size()));
+    }
+  }
+
+  // check tileSizes is full on or multiple of `inner_tile_size`
+  for (auto [tile, innerTile] :
+       llvm::zip_equal(targetTileSizes, targetInnerTileSizes)) {
+    if (isEqualConstantIntOrValue(tile, innerTile))
+      continue;
+    FailureOr<int64_t> cstSize = ValueBoundsConstraintSet::computeConstantBound(
+        presburger::BoundType::UB,
+        getValueOrCreateConstantIndexOp(rewriter, op->getLoc(), tile),
+        /*dim=*/std::nullopt,
+        /*stopCondition=*/nullptr, /*closedUB=*/true);
+    std::optional<int64_t> cstInnerSize = getConstantIntValue(innerTile);
+    if (!failed(cstSize) && cstInnerSize) {
+      if (*cstSize % *cstInnerSize == 0)
+        continue;
+    }
+    return failure();
+  }
+  return success();
+}
+
+template <typename T>
+void FusionAnchorBase<T>::appendCandidateWithVerifyTileSizes(
+    RewriterBase &rewriter,
+    ArrayRef<OffsetSizeAndStrideOpInterface> candidateList) {
+  for (auto candidate : candidateList) {
+    if (succeeded(verifyTilableOpTileSizesOnDimAndTileMap<T>(
+            rewriter, fusableOp, candidate.getMixedSizes()))) {
+      candidateSliceOpList.push_back(cast<T>(candidate));
+    }
+  }
+}
+
+template <typename T>
+FailureOr<T>
+FusionAnchorBase<T>::selectCandidateByCostModel(RewriterBase &rewriter,
+                                                Location loc, SystemDesc desc) {
+  if (candidateSliceOpList.empty())
+    return failure();
+  /// TODO: use cost model
+  if (std::is_same<T, tensor::ExtractSliceOp>::value) {
+    return cast<T>(candidateSliceOpList.front());
+  } else {
+    return cast<T>(candidateSliceOpList.back());
+  }
+}
+
+// Target at tensor.extract_slice
+class ProducerFusionAnchor : public FusionAnchorBase<tensor::ExtractSliceOp> {
+public:
+  ProducerFusionAnchor(
+      RewriterBase &rewriter, Operation *producerOp, OpResult producerValue,
+      SmallVector<tensor::ExtractSliceOp> candidateExtractSliceOpList)
+      : FusionAnchorBase<tensor::ExtractSliceOp>(producerOp) {
+    auto candidateList = llvm::map_to_vector(
+        candidateExtractSliceOpList,
+        [](tensor::ExtractSliceOp sliceOp) -> OffsetSizeAndStrideOpInterface {
+          return sliceOp;
+        });
+    if (auto linalgOp = dyn_cast<linalg::LinalgOp>(producerOp)) {
+      appendCandidateWithVerifyTileSizes(
+          rewriter, candidateList,
+          linalgOp.getIndexingMapMatchingResult(producerValue));
+    } else if (isa<tensor::PackOp, tensor::UnPackOp, tensor::PadOp>(
```

Comment:
Does `tensor::ConcatOp tensor::bitcastOp` need to process in here?

Reply:
This pass only focus tilable op as its name. Currently, only `pad/pack/unpack` operation implement `TensorTilingInterface`. 

Besides:
1. We will use [`linalg.copy`](https://mlir.llvm.org/docs/Dialects/Linalg/#linalgcopy-linalgcopyop) rather than `tensor::bitcastOp` to represent elementwise cast semantic.
2. The result of `tensor::empty` is usually used as `init` of Dps Op. Currently, if one DPS is fused into loop, we need to append it `init` into `iter_arg` of loop. As the result, it seems unnecessary to tile an `empty` op.
3. As for `concat`, we can enhance it in the future by concrete use case. BTW, it can also be optimized it via `bufferization`.